corehq/apps/ota/models.py

```diff
@@ -131,3 +132,85 @@ def _get_next(cls, domain, bucket_id):
+
+
+class MobileRecoveryMeasure(models.Model):
+    """
+    Model representing a method of recovering from a fatal error on mobile.
+    """
+    MEASURES = (
+        Measure('app_reinstall', "Reinstall App, OTA",
+                "Reinstall the current CommCare app by triggering an OTA install"),
+        Measure('app_reinstall_local', "Reinstall App, Local",
+                "Reinstall the current CommCare app (by triggering an offline install "
+                "with a default .ccz that's already on the device, and then doing an "
+                "OTA update from there)"),
+        Measure('app_update', "Update App",
+                "Update the current CommCare app"),
+        Measure('clear_data', "Clear User Data",
+                "Clear data for the app's last logged in user"),
+        Measure('cc_reinstall', "CC Reinstall Needed",
+                "Notify the user that CommCare needs to be reinstalled"),
+        Measure('cc_update', "CC Update Needed",
+                "Notify the user that CommCare needs to be updated"),
+    )
+    measure = models.CharField(
+        max_length=255,
+        choices=[(m.slug, m.name) for m in MEASURES],
+        help_text="<br/>".join(
+            "<strong>{}:</strong> {}".format(m.name, m.description)
+            for m in MEASURES
+        )
+    )
+
+    domain = models.CharField(max_length=255)
+    app_id = models.CharField(max_length=50)
+    sequence_number = models.IntegerField(editable=False)
+
+    cc_all_versions = models.BooleanField(
+        verbose_name="All CommCare Versions", default=True)
+    cc_version_min = models.CharField(
+        verbose_name="Min CommCare Version", max_length=255, blank=True)
+    cc_version_max = models.CharField(
+        verbose_name="Max CommCare Version", max_length=255, blank=True)
+
+    app_all_versions = models.BooleanField(
+        verbose_name="All App Versions", default=True)
+    app_version_min = models.CharField(
+        verbose_name="Min App Version", max_length=255, blank=True)
+    app_version_max = models.CharField(
+        verbose_name="Max App Version", max_length=255, blank=True)
+
+    created_on = models.DateTimeField(auto_now_add=True)
+    username = models.CharField(max_length=255, editable=False)
+    notes = models.TextField(blank=True)
+
+    class Meta(object):
+        unique_together = ('domain', 'app_id', 'sequence_number',)
+
+    def set_sequence_number(self):
+        # It's unlikely there will ever be concurrency issues with this
+        # so it seems reasonable to let the db constraint prevent conflicts
+        current_max = (MobileRecoveryMeasure.objects
+                       .filter(domain=self.domain,
+                               app_id=self.app_id)
+                       .order_by('-sequence_number')
+                       .values_list('sequence_number', flat=True)
+                       .first()) or 0
+        self.sequence_number = current_max + 1
```

Comment:
As noted in the spec the requirement for this to be sequential seems odd to me. I think it would make more sense to just use the primary key of the model and for mobile to treat the 'sequence_number' as an opaque ID.

@amstone326 @esoergel

Reply:
Commenting here instead of on the question on the spec-- @snopoke Is your concern just the cost of computing the next sequence number for a record? From the mobile standpoint, it's simpler and faster to be able to keep track of just the highest version so far, so I suppose I was prioritizing that. And since the expected load for these is always going to be very low, I don't think it's going to make a measurable difference on either side of the equation, but I'm fine with changing it if this seems substantially worse from the perspective of the django model. @esoergel, let me know if you change this so that I can make mobile changes accordingly.

Reply:
I think if we make the constraint that they always go in increasing order you would still be able to do that. I just think enforcing that they are incremental is unnecessary.

Reply:
I can make that change - Aliza, django primary keys are incrementally increasing integers, so each new key will have a higher number than those preceding it.  The only difference is that that pool of numbers would be irrespective of domain and app ID, so if you have 6 existing measures, then make a measure for a new domain, it would have a sequence number of 7.